### PR TITLE
 ENH: Build VXL with WINDOWS_EXPORT_ALL_SYMBOLS & BUILD_SHARED

### PR DESCRIPTION
**Pull request by hjmjohnson**
Windows linkage issues are present with these two options set.

'identifier' : member of dll interface class may not be declared with dll interface
You can declare a whole class, or certain members of a non-DLL interface class, with DLL interface. You cannot declare a class with DLL interface and then declare a member of that class with DLL interface.

'identifier' : definition of dllimport function not allowed
Data, static data members, and functions can be declared as dllimports but not defined as dllimports.
To fix this issue, remove the __declspec(dllimport) specifier from the definition of the function.

### CODE REVIEW

**Comment by hjmjohnson** [comment]
@dzenanz  THIS DOES NOT WORK YET!!!

I started down the path and fixed some problems.  I'm stuck now.  I could use your help to identify how to fix the next set of link errors.

Hasn

File: core/vnl/Templates/vnl_vector_ref+double-.cxx [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#include <vnl/vnl_vector_ref.hxx>
+VNL_VECTOR_REF_INSTANTIATE(double);
+VNL_VECTOR_REF_INSTANTIATE(double);
```
Comment (dzenanz) [inline core/vnl/Templates/vnl_vector_ref+double-.cxx]: Are you instantiating twice on purpose?

File: core/vnl/Templates/vnl_vector_ref+float-.cxx [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#include <vnl/vnl_vector_ref.hxx>
+VNL_VECTOR_REF_INSTANTIATE(float);
+VNL_VECTOR_REF_INSTANTIATE(float);
```
Comment (dzenanz) [inline core/vnl/Templates/vnl_vector_ref+float-.cxx]: Are you instantiating twice on purpose?

File: core/vnl/vnl_vector_ref.h [hunk 3]
```diff
@@ -24,28 +24,20 @@ template <class T>
...
   // Do \e not call anything else than the default constructor of vnl_vector<T>
-  vnl_vector_ref(unsigned n, T *space) : vnl_vector<T>() {
-    Base::data = space;
-    Base::num_elmts = n;
-  }
+  vnl_vector_ref(unsigned n, T *space);
 
   //: Copy constructor
   // Do \e not call anything else than the default constructor of vnl_vector<T>
   // (That is why the default copy constructor is \e not good.)
```
Comment (dzenanz) [inline core/vnl/vnl_vector_ref.h]: What does this backslash-e do in the comment? And it appears on the previous line too.

File: core/vnl/Templates/vnl_vector_ref+double-.cxx (hunk 1 above)
Comment (hjmjohnson) [reply core/vnl/Templates/vnl_vector_ref+double-.cxx]: That was an error.

File: core/vnl/Templates/vnl_vector_ref+float-.cxx (hunk 2 above)
Comment (hjmjohnson) [reply core/vnl/Templates/vnl_vector_ref+float-.cxx]: an error.

**Comment by hjmjohnson** [comment]
@dzenanz @thewtex This is now ready to review.

Windows build with export all symbols and build shared:
See https://open.cdash.org/buildSummary.php?buildid=5627435

Mac build with build shared:
See https://open.cdash.org/buildSummary.php?buildid=5627440

File: core/vnl/Templates/vnl_matrix_fixed+complex_float.allsizes-.cxx [hunk 4]
```diff
@@ -1,2 +1,5 @@
 #include <vnl/vnl_matrix_fixed.hxx>
-VNL_MATRIX_FIXED_INSTANTIATE(std::complex<float>,2,2);
+VNL_MATRIX_FIXED_INSTANTIATE(std::complex<float>, 2, 2);
+VNL_MATRIX_FIXED_INSTANTIATE(std::complex<float>, 3, 3);
+VNL_MATRIX_FIXED_INSTANTIATE(std::complex<float>, 4, 4);
+VNL_MATRIX_FIXED_INSTANTIATE(std::complex<float>, 5, 5);
```
Comment (dzenanz) [inline core/vnl/Templates/vnl_matrix_fixed+complex_float.allsizes-.cxx]: No newline.

Comment (hjmjohnson) [reply core/vnl/Templates/vnl_matrix_fixed+complex_float.allsizes-.cxx]: ?? I don't understand.

**Comment by hjmjohnson** [comment]
@dzenanz   The last patch contains the explicit instantiations that will all ITK to have NO extra instantiations in the VNLInstantiation directory.  That will make debugging easier.

32>Project not selected to build for this solution configuration 
========== Build: 26 succeeded, 0 failed, 2 up-to-date, 6 skipped ==========

File: core/vnl/Templates/vnl_matrix_fixed+complex_float.allsizes-.cxx (hunk 4 above)
Comment (hjmjohnson) [reply core/vnl/Templates/vnl_matrix_fixed+complex_float.allsizes-.cxx]: Oh.  I fixed that in the next commit.

**Comment by thewtex** [comment]
Fantastic!!!

**Pull request merged**